### ISSUE TITLE
This adapter is not PSR-7 compliant

### ISSUE THREAD

**Issue by stof** [Sam-Burns/psr7-symfony-httpfoundation#1]
An important PSR-7 requirement is the immutability of the Request and Response object. Anything implemented as a wrapper around a mutable HttpFoundation Request/Response **cannot** be immutable (the HttpFoundation Request can be mutated externally)

**Comment by Sam-Burns** [comment]
@stof Why do you think the HttpFoundation Request can be mutated externally?  If it is a private property of a PSR-7 Request object, that shouldn't be a problem.

Perhaps your concern is that a Symfony Request object could be passed in via constructor injection, and then mutated externally?  In that case, simply cloning the Symfony Request in the constructor and then saving it as a private class property would solve that problem.

If your concern is, for example, that the `HttpFoundation\Request::getHost()` method or the `getPort()` method contain if statements whose conditions refer to static properties on the class, we should be mindful of the fact that the implementation here will need its own Uri class anyway, wrapping PHP's native `parse_uri()` function.  This is because, to cut a long story short, URI's are strings not objects in Symfony.  This means that, despite being predominantly a wrapper for Symfony classes, the library will have its own (immutable) Uri class, which can be relied upon to provide `getPort()` and `getHost()` methods invokable from the Request class without compromising immutability.

Even implementing `Request::getHost()` with code like:

``` php
$wrappedSymfonyReqeust->setTrustedHeaders(array());
return $wrappedSymfonyRequest->getHost();
```

basically provides a Request object where no use case can materially mutate the wrapped object, if need be.

Your feedback on this issue is very welcome, @stof, and thank you for getting involved :-)  Perhaps once more of the implementation is in place, you could help out by trying to provide use cases where the wrapped object is mutated in a way that affects behaviour?  I am still hopeful that there won't be any.

**Comment by ciaranmcnulty** [comment]
@stof It looks like @Sam-Burns has added a `clone` to the Request constructor. From looking at Symfony\Request which implements a `__clone` I think this is sufficient to guard against mutability?

**Comment by Sam-Burns** [comment]
@ciaranmcnulty's [PHPUnit test](http://github.com/Sam-Burns/psr7-symfony-httpfoundation/commit/7260e1aea5a0cde17a8d52b42818dc0b245be734) demonstrates immutability of headers in the wrapper.  Will be on the lookout for any other places where mutation could occur.

**Comment by Sam-Burns** [comment]
Closing the ticket, as the class is immutable now.  I will continue to monitor the situation as development progresses.
